src/utils/pluginDriver.ts

```diff
@@ -0,0 +1,216 @@
+import Graph from '../Graph';
+import {
+	InputOptions,
+	Plugin,
+	PluginContext,
+	RollupError,
+	RollupWarning,
+	Watcher
+} from '../rollup/types';
+import { createAssetPluginHooks, EmitAsset } from './assetHooks';
+import { getRollupDefaultPlugin } from './default-plugin';
+import error from './error';
+
+export interface PluginDriver {
+	emitAsset: EmitAsset;
+	getAssetFileName(assetId: string): string;
+	hookSeq(hook: string, args?: any[], context?: HookContext): Promise<void>;
+	hookFirst<T = any>(hook: string, args?: any[], hookContext?: HookContext): Promise<T>;
+	hookParallel(hook: string, args?: any[], hookContext?: HookContext): Promise<void>;
+	hookReduceArg0<R = any, T = any>(
+		hook: string,
+		args: any[],
+		reduce: Reduce<R, T>,
+		hookContext?: HookContext
+	): Promise<T>;
+	hookReduceValue<R = any, T = any>(
+		hook: string,
+		value: T | Promise<T>,
+		args: any[],
+		reduce: Reduce<R, T>,
+		hookContext?: HookContext
+	): Promise<T>;
+	hasLoadersOrTransforms: boolean;
+}
+
+export type Reduce<R = any, T = any> = (reduction: T, result: R, plugin: Plugin) => T;
+export type HookContext = (context: PluginContext, plugin?: Plugin) => PluginContext;
+
+export function createPluginDriver(
+	graph: Graph,
+	options: InputOptions,
+	watcher?: Watcher
+): PluginDriver {
+	const plugins = [...(options.plugins || []), getRollupDefaultPlugin(options)];
+	const { emitAsset, getAssetFileName, setAssetSource } = createAssetPluginHooks(graph.assetsById);
+	const existingPluginNames: string[] = [];
+
+	let hasLoadersOrTransforms = false;
+
+	const pluginContexts = plugins.map(plugin => {
+		if (typeof plugin.name === 'string') {
+			if (existingPluginNames.indexOf(plugin.name) !== -1)
+				error({
+					code: 'DUPLICATE_PLUGIN_NAME',
+					message: `The plugin name ${
```

Comment:
while this is true - plugins should have unique names, I'm not sure if you should fail the build in this case - this might cause trouble for rollup users because of a rollup plugin author's mistake

also - it seems reasonable to i.e. use `rollup-plugin-babel` multiple times (with different babel configs on different paths - using exclude/include options)